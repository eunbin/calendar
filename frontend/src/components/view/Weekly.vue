<template>
  <table class="week-view">
    <thead>
      <tr>
        <th />
        <th
          v-for="(day, index) in dayOfWeek"
          :key="index"
        >
          <span class="day">{{ day.format(dateFormat.DATE_AND_DAY) }}</span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="(hour, index) in hours"
        :key="index"
      >
        <td class="hour">
          <span>{{ hour.format(dateFormat.HOUR_AND_MIN) }}</span>
        </td>
        <td
          v-for="(day, i) in dayOfWeek"
          :key="i"
          class="hour"
          @dragover.prevent
          @dragenter.prevent
          @drop.prevent="onDrop(day, hour)"
          @click.stop="selectHour(day, hour)"
        >
          <dl class="event-list">
            <dt
              v-for="(event, j) in getEventsByMap(day, hour)"
              :key="j"
              :class="{ 'selected': event.selected }"
              class="event"
              draggable="true"
              @dragstart="onDragStart(event)"
              @click.stop="selectEvent(event)"
            >
              <span>{{ event.title }}</span>
            </dt>
          </dl>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
import moment from 'moment'
import dateMixin from '@/mixins/date'

export default {
  name: 'Weekly',
  mixins: [dateMixin],
  props: {
    currentDate: {
      type: Object,
      default: null
    },
    eventsMap: {
      type: Object,
      default: () => {}
    }
  },
  computed: {
    dayOfWeek () {
      const startOfWeek = this.currentDate.clone().startOf('week').subtract(1, 'd')
      return moment.weekdays().map(() => startOfWeek.add(1, 'd').clone())
    },
    hours () {
      return Array(24).fill('').map((_, i) => moment(i, 'hh'))
    }
  },
  methods: {
    getEventsByMap (day, hour) {
      const events = this.eventsMap[day.format(this.dateFormat.DATE)]
      if (events) {
        return events.filter(event => moment(event.start.dateTime).hours() === hour.hours())
      }
      return []
    },
    onDrop (date, hour) {
      const event = this.draggingEvent
      if (event) {
        const { start, end } = event
        const newDate = hour ? date.clone().hours(hour.hours()) : date.clone().hours(moment(start.dateTime).hours())
        const payload = {
          ...event,
          start: {
            date: moment(start.date).month(newDate.month()).date(newDate.date()).format(this.dateFormat.DATE),
            dateTime: moment(start.dateTime).month(newDate.month()).date(newDate.date()).hour(newDate.hours()).format(this.dateFormat.DATE_TIME)
          },
          end: {
            date: moment(end.date).month(newDate.month()).date(newDate.date()).format(this.dateFormat.DATE),
            dateTime: moment(end.dateTime).month(newDate.month()).date(newDate.date()).hour(newDate.hours() + 1).format(this.dateFormat.DATE_TIME)
          }
        }
        this.$emit('event-moved', payload)
      }
      this.draggingEvent = null
    },
    selectHour (day, hour) {
      this.$emit('hour-selected', day, hour)
    },
    onDragStart (event) {
      this.draggingEvent = event
    },
    // TODO: event delegation
    selectEvent (event) {
      this.$emit('event-selected', event)
    }
  }
}
</script>

<style lang="scss" scoped>
  table.week-view {
    width: 100%;
    height: 100%;
    border-spacing: 0;
    border-collapse: separate;
    table-layout: fixed;
    thead {
      tr {
        th {
          background: var(--color-primary);
          color: var(--color-text);
          padding: 0.5em;
          overflow: hidden;
          &:first-child {
            width: 50px;
            border-radius: 3px 0 0 0;
          }
          &:last-child {
            border-radius: 0 3px 0 0;
          }
          .day {
            font-size: 1.2em;
            border-radius: 50%;
            line-height: 1.8;
            &.active {
              background: var(--color-gray);
              color: var(--color-primary);
            }
          }
          .short {
            display: none;
          }
          i {
            vertical-align: middle;
            font-size: 2em;
          }
        }
      }
    }

    tbody {
      tr {
        background: var(--color-gray);
        height: calc(100%/24);
        &:nth-child(odd) {
          background: var(--color-gray);
        }
        &:nth-child(4n+0) {
          td {
            border-bottom: 1px solid var(--color-primary);
          }
        }
        td {
          text-align: center;
          vertical-align: middle;
          border-left: 1px solid var(--color-dark-gray);
          border-bottom: 1px solid var(--color-dark-gray);
          position: relative;
          height: 32px;
          &:last-child {
            border-right: 1px solid var(--color-dark-gray);
          }
          &.hour {
            font-size: 1.2em;
            padding: 0;
            color: var(--color-text);
            background: #fff;
            border-bottom: 1px solid var(--color-dark-gray);
            border-collapse: separate;
            min-width: 100px;
            span {
              display: block;
            }
          }
        }
      }
    }
  }
</style>
