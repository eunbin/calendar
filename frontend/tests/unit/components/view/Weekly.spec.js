import Weekly from '@/components/view/Weekly.vue'
import dateMixin from '@/mixins/date'
import { createLocalVue, mount } from '@vue/test-utils'
import { currentDate } from '../../test-data'
import moment from 'moment'
import Vuex from 'vuex'

const createOption = () => {
  return {
    mixins: [dateMixin],
    propsData: {
      currentDate: moment(currentDate),
      today: moment(),
      eventsMap: {}
    }
  }
}

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Weekly.vue', () => {
  let store
  let getters = {
    getEventsByDayAndHour: () => () => []
  }

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        events: {}
      },
      getters
    })
  })
  it('컴포넌트가 정상적으로 동작되면 days-of-week 컬럼이 생성된다.', async () => {
    const wrapper = mount(Weekly, Object.assign(createOption(), {
      localVue,
      store
    }))
    const columnsWrapper = wrapper.findAll('th')
    const DAYS = 7
    const TIME_COLUMN = 1
    expect(columnsWrapper.length).toBe(DAYS + TIME_COLUMN)
  })
})
