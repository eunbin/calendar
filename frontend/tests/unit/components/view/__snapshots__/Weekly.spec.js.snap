// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weekly.vue snapshot check 1`] = `
<table
  class="week-view"
  today="Tue Feb 18 2020 02:34:07 GMT+0900"
>
  <thead>
    <tr>
      <th />
       
      <th>
        <span
          class="day"
        >
          16 (Su)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          17 (Mo)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          18 (Tu)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          19 (We)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          20 (Th)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          21 (Fr)
        </span>
      </th>
      <th>
        <span
          class="day"
        >
          22 (Sa)
        </span>
      </th>
    </tr>
  </thead>
   
  <tbody>
    <tr>
      <td
        class="hour"
      >
        <span>
          00:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          01:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          02:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          03:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          04:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          05:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          06:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          07:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          08:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          09:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          10:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          11:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          12:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          13:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          14:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          15:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          16:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          17:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          18:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          19:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          20:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          21:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          22:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
    <tr>
      <td
        class="hour"
      >
        <span>
          23:00
        </span>
      </td>
       
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
      <td
        class="hour"
      >
        <dl
          class="event-list"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
